### PR TITLE
trainer: add Megatron-Core Tensor Parallelism user guide

### PR DESCRIPTION
**Pull request by XploY04**
### Description of Changes

This PR adds a Megatron user guide describing how to train models using
Megatron-Core with Tensor Parallelism on Kubeflow Trainer.

The guide covers:
- Megatron-Core overview (TP, PP, DP)
- Distributed environment setup with `torch-distributed` runtime
- Full GPT training example with Tensor Parallelism
- Increasing `/dev/shm` for NCCL in Kubernetes
- Multi-node vs multi-GPU configuration (including GPU time-slicing)

### Related Issues

Related: https://github.com/kubeflow/trainer/issues/3178
Related: https://github.com/kubeflow/trainer/pull/3201

### Checklist

- [x] You have [signed off your commits](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits)
- [x] Ensure you follow best practices from our [contributing guide](https://github.com/kubeflow/website/blob/master/content/en/docs/about/contributing.md).
- [ ] (for big changes) I will post screenshots of the changes in a PR comment

### CODE REVIEW

**Comment by XploY04** [comment]
@andreyvelich please review

**Review by Arhell — COMMENTED** [review]
/ok-to-test

File: content/en/docs/components/trainer/user-guides/megatron.md [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+        module=gpt_model,
+    )
+
+    optim = Adam(gpt_model.parameters())
+
+    # Step 3: Prepare a mock dataset (no real data download needed).
+    # Install build tools for Megatron's C++ dataset helpers.
+    import subprocess as _sp
+    _sp.run(["apt-get", "update", "-qq"], capture_output=True)
+    _sp.run(["apt-get", "install", "-y", "-qq", "make", "g++"], capture_output=True)
```
Comment (andreyvelich) [inline content/en/docs/components/trainer/user-guides/megatron.md:174]: Could you also explain this part in the Megatron Distributed Environment section?
E.g. you can say that the `g++` compiler must be installed in the Runtime image to run Megatron.

File: content/en/docs/components/trainer/user-guides/megatron.md [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    )
+    for _f in ["Makefile", "helpers.cpp"]:
+        _urlreq.urlretrieve(_base_url + _f, os.path.join(_datasets_dir, _f))
+
+    if torch.distributed.is_available() and torch.distributed.is_initialized():
+        if torch.distributed.get_rank() == 0:
+            compile_helpers()
+        torch.distributed.barrier()
+    else:
+        compile_helpers()
```
Comment (andreyvelich) [inline content/en/docs/components/trainer/user-guides/megatron.md]: Do we need this?

File: content/en/docs/components/trainer/user-guides/megatron.md [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+  Megatron-Core includes `DistributedDataParallel` for gradient synchronization.
+
+Since Megatron-Core uses `torchrun` as its distributed launcher, it works natively with the
+existing `torch-distributed` ClusterTrainingRuntime. No dedicated Megatron runtime is needed.
+
+## Megatron Distributed Environment
+
+Kubeflow Trainer uses the `torch-distributed` runtime to launch Megatron-Core training with
+`torchrun` on every node. The standard PyTorch distributed environment variables are available
+in your training function:
```
Comment (andreyvelich) [inline content/en/docs/components/trainer/user-guides/megatron.md:45]: What about `/dev/shm` memory problem that we saw in the tests?

File: content/en/docs/components/trainer/user-guides/megatron.md [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+        compile_helpers()
+
+    config = GPTDatasetConfig(
+        random_seed=0,
+        sequence_length=_SEQUENCE_LENGTH,
+        reset_position_ids=False,
+        reset_attention_mask=False,
+        eod_mask_loss=False,
+        tokenizer=MegatronTokenizer.from_pretrained(
+            metadata_path={"library": "null"},
```
Comment (andreyvelich) [inline content/en/docs/components/trainer/user-guides/megatron.md]: Should this be `null-text` ?
https://github.com/kubeflow/trainer/pull/3434

File: content/en/docs/components/trainer/user-guides/megatron.md [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+
+## Prerequisites
+
+Before exploring this guide, make sure to follow [the Getting Started guide](/docs/components/trainer/getting-started/)
+to understand the basics of Kubeflow Trainer.
+
+Your cluster needs at least **2 NVIDIA GPUs** (across one or more nodes). Megatron-Core requires
+CUDA and uses the NCCL backend for distributed communication.
+
+## Megatron-Core Overview
```
Comment (andreyvelich) [inline content/en/docs/components/trainer/user-guides/megatron.md:18]: Maybe you can also briefly say about Megatron Bridge which allows to convert model into Megatron-compatible format for TP and PP?

File: content/en/docs/components/trainer/user-guides/megatron.md [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+        },
+        packages_to_install=["megatron-core", "pybind11"],
+        env={"TP_SIZE": str(tensor_model_parallel_size)},
+    ),
+)
+```
+
+{{% alert title="Note" color="info" %}}
+The `TP_SIZE` environment variable must match the total number of GPUs across all nodes
+(`num_nodes * gpu`). If you set `num_nodes=2` and `gpu=1`, then `TP_SIZE` should be 2.
```
Comment (andreyvelich) [inline content/en/docs/components/trainer/user-guides/megatron.md]: Try to avoid using code snippets in the Notes, as they tend to render very blurry.

```suggestion
The TP_SIZE environment variable must match the total number of GPUs across all nodes
(`num_nodes * gpu`). If you set `num_nodes=2` and `gpu=1`, then `TP_SIZE` should be 2.
```

File: content/en/docs/components/trainer/user-guides/megatron.md [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+
+## Increasing `/dev/shm` for NCCL
+
+NCCL allocates shared memory in `/dev/shm` for its proxy service (roughly 33 MB per communicator).
+Kubernetes pods get 64 MB of `/dev/shm` by default, which is not enough for Megatron-Core because
+it creates multiple NCCL communicators (one per parallel group: TP, DP, etc.). When `/dev/shm`
+fills up, NCCL operations fail silently or crash.
+
+The fix is to mount `/dev/shm` as a memory-backed `emptyDir` volume in the ClusterTrainingRuntime.
+If your cluster administrator has not already configured this, you can patch the runtime:
```
Comment (andreyvelich) [inline content/en/docs/components/trainer/user-guides/megatron.md]: Please can you also add example with TrainJob and RuntimePatches?

**Review by andreyvelich — COMMENTED** [review]
Overall, looks great, thanks @XploY04!
cc @astefanutti @tenzen-y @kuizhiqing @jaiakash

File: content/en/docs/components/trainer/user-guides/megatron.md (hunk 3 above)
Comment (andreyvelich) [reply content/en/docs/components/trainer/user-guides/megatron.md:45]: Found it in the appropriate sections, thanks!

File: content/en/docs/components/trainer/user-guides/megatron.md [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+        ddp_config=ddp_config,
+        module=gpt_model,
+    )
+
+    optim = Adam(gpt_model.parameters())
+
+    # Step 3: Prepare a mock dataset (no real data download needed).
+    # Install build tools for Megatron's C++ dataset helpers.
+    import subprocess as _sp
+    _sp.run(["apt-get", "update", "-qq"], capture_output=True)
```
Comment (astefanutti) [inline content/en/docs/components/trainer/user-guides/megatron.md:173]: That should probably belong to the training runtime and not the training function.
We may want to add it to the training runtime image.

File: content/en/docs/components/trainer/user-guides/megatron.md [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+training with frameworks that create several NCCL communicators.
+
+## Multi-Node vs Multi-GPU Configuration
+
+When configuring your TrainJob, the relationship between `num_nodes`, `gpu` per node, and
+`TP_SIZE` matters:
+
+| Configuration | `num_nodes` | `gpu` | `TP_SIZE` | Pods | How it works |
+|---|---|---|---|---|---|
+| Multi-GPU single node | 1 | 2 | 2 | 1 | 2 workers in 1 pod, each gets a CUDA device |
```
Comment (astefanutti) [inline content/en/docs/components/trainer/user-guides/megatron.md:378]: Would it be useful to show how to have those two pods scheduled on the same node, e.g. when using NVLink, as TP is network intensive.

**Review by astefanutti — COMMENTED** [review]
Thanks @XploY04!

I've left few small comments, otherwise LGTM.

/lgtm

**Comment by andreyvelich** [comment]
@XploY04 Did you get a chance to address feedback?

**Comment by XploY04** [comment]
@andreyvelich I was busy with some work, I will do that today.

File: content/en/docs/components/trainer/user-guides/megatron.md (hunk 5 above)
Comment (XploY04) [reply content/en/docs/components/trainer/user-guides/megatron.md:18]: Added a paragraph at the end of the Megatron-Core Overview section pointing at Megatron-Bridge and clarifying that this guide sticks to Megatron-Core directly with a GPT model built from scratch. Let me know if you want it expanded further.

File: content/en/docs/components/trainer/user-guides/megatron.md (hunk 1 above)
Comment (XploY04) [reply content/en/docs/components/trainer/user-guides/megatron.md:174]: Added a paragraph to the Megatron Distributed Environment section explaining that the dataset helpers compile from C++ at first use and that the training container needs make and g++. Also noted that pytorch/pytorch:*-runtime does not ship these tools, so the example installs them via apt-get and production users should bake them into a custom runtime image. The same workaround note appears right below the code block

File: content/en/docs/components/trainer/user-guides/megatron.md (hunk 2 above)
Comment (XploY04) [reply content/en/docs/components/trainer/user-guides/megatron.md]: Removed. Inside a TrainJob torch.distributed.init_process_group() always runs a few lines above, so is_initialized() is always true and the else branch was dead code. The snippet is now just:                                
```
if torch.distributed.get_rank() == 0:
     compile_helpers()                                   
torch.distributed.barrier()  
```

File: content/en/docs/components/trainer/user-guides/megatron.md (hunk 4 above)
Comment (XploY04) [reply content/en/docs/components/trainer/user-guides/megatron.md]: Yes, changed to `null-text`

File: content/en/docs/components/trainer/user-guides/megatron.md (hunk 7 above)
Comment (XploY04) [reply content/en/docs/components/trainer/user-guides/megatron.md]: Added a second YAML block under Increasing /dev/shm for NCCL showing the same mount applied per-TrainJob via runtimePatches. The section now has two clearly labeled options: Option 1 for cluster administrators patching the ClusterTrainingRuntime once, and Option 2 for TrainJob authors who can't modify the runtime. Also fixed the container name from trainer to node in the existing example to match the torch-distributed runtime.

File: content/en/docs/components/trainer/user-guides/megatron.md (hunk 9 above)
Comment (XploY04) [reply content/en/docs/components/trainer/user-guides/megatron.md:378]: Added a new subsection Co-Scheduling Multi-Node Pods on the Same Physical Node under Multi-Node vs Multi-GPU Configuration. It includes a short explanation of why this matters for TP and a runtimePatches example using podAffinity with the jobset.sigs.k8s.io/jobset-name label and topologyKey: kubernetes.io/hostname.

File: content/en/docs/components/trainer/user-guides/megatron.md (hunk 8 above)
Comment (XploY04) [reply content/en/docs/components/trainer/user-guides/megatron.md:173]: Agreed. For now I've added a Note right after the code block flagging this as a workaround and recommending the runtime image as the proper fix, plus a matching heads-up in the Megatron Distributed Environment section. I'd prefer to track the actual runtime image change as a separate issue/PR against kubeflow/trainer so this docs PR stays focused. Happy to open one if you'd like.

**Review by andreyvelich — COMMENTED** [review]
Thanks for this @XploY04!
/lgtm
/approve
